File: api/src/context.rs
use mongodb::Database;
use rocket::{
    http::Status,
    outcome::try_outcome,
    request::{FromRequest, Outcome},
    Request, State,
};

use crate::{
    models::{BaseModel, Questions, QuizTemplates, User, Users},
    session::SessionStorage,
};

pub struct Context {
    pub questions: Questions,
    pub quiz_templates: QuizTemplates,
    pub users: Users,
    pub user: Option<User>,
}

impl juniper::Context for Context {}

#[rocket::async_trait]
impl<'r> FromRequest<'r> for Context {
    type Error = ();

    async fn from_request(request: &'r Request<'_>) -> Outcome<Self, Self::Error> {
        let db: &State<Database> = try_outcome!(request.guard::<&State<mongodb::Database>>().await);

        let users = Users::new(&db);

        let mut session_storage: SessionStorage =
            try_outcome!(request.guard::<SessionStorage>().await);

        let user_id = match session_storage.get_session(request.cookies()).await {
            Ok(user_id) => user_id,
            Err(e) => {
                eprintln!("Error getting session: {:?}", e);

                return Outcome::Error((Status::InternalServerError, ()));
            }
        };

        let user = match user_id {
            Some(id) => match users.read_by_id(&id).await {
                Ok(user) => user,
                Err(e) => {
                    eprintln!("Error reading user: {:?}", e);

                    return Outcome::Error((Status::InternalServerError, ()));
                }
            },
            None => None,
        };

        Outcome::Success(Context {
            questions: Questions::new(&db),
            quiz_templates: QuizTemplates::new(&db),
            user,
            users,
        })
    }
}

File: api/app/context.py
from dataclasses import dataclass
from typing import Optional
from flask import g, request

from .models import SESSION_COOKIE_NAME, Questions, QuizTemplates, Sessions, UserModel, Users

@dataclass
class Context:
    current_user: Optional[UserModel]
    questions: Questions
    quiz_templates: QuizTemplates
    sessions: Sessions
    users: Users

def get_context() -> Context:
    if "context" not in g:
        sessions = Sessions()
        users = Users()

        user = None

        if SESSION_COOKIE_NAME in request.cookies:
            session_id = request.cookies[SESSION_COOKIE_NAME]
            user_id = sessions.get(session_id)["user_id"]
            try:
                user = Users().get_by_id(user_id)
            except ValueError:
                pass

        g.context = Context(
            current_user=user,
            questions=Questions(),
            quiz_templates=QuizTemplates(),
            sessions=sessions,
            users=users,
        )
    
    return g.context